Objects/TI4_BuildArea.ttslua

```diff
@@ -600,6 +618,23 @@ function getProduceConsumeItems()
         end
     end
 
+    if result.aiDevelopmentAlgorithm then
+        -- Detect unit upgrades (AI Development Algorithm)
+        local unitUpgrades = {}
+        for _, object in ipairs(getAllObjects()) do
+            if not inHandGuidSet[object.getGUID()] and not object.is_face_down and object.tag == 'Card'
+                and playerColor and _zoneHelper.zoneFromPosition(object.getPosition()) == playerColor then
```

Comment:
I don't know if calling a different object is expensive but would be nice to either delay calling zoneHelper until it is more certain this is an object of interest, or my usual style is to build a map from guidToPosition then use the zoneHelper.zonesFromPositions method to do them all with a single call.

Reply:
Thinking about this a bit more, the unit helper function could return all upgrades.